runtime-sdk/src/modules/consensus_accounts/mod.rs

```diff
@@ -0,0 +1,383 @@
+            &amount,
+            MessageEventHookInvocation::new(
+                CONSENSUS_WITHDRAW_HANDLER.to_string(),
+                types::ConsensusWithdrawContext {
+                    address: from,
+                    amount: amount.clone(),
+                },
+            ),
+        )?;
+
+        Ok(())
+    }
+
+    fn withdraw<C: Context>(
+        ctx: &mut C,
+        to: Address,
+        amount: token::BaseUnits,
+    ) -> Result<(), Error> {
+        if ctx.is_check_only() {
+            return Ok(());
+        }
+
+        // Check internal store if account has enough balance to withdraw.
+        let balances =
+            Accounts::get_balances(ctx.runtime_state(), to).map_err(|_| Error::InvalidArgument)?;
+        let balance = balances
+            .balances
+            .get(amount.denomination())
+            .ok_or(Error::InvalidArgument)?;
+        if balance < amount.amount() {
+            return Err(Error::InsufficientWithdrawBalance);
+        }
+
+        // Transfer out of runtime account and update the account state if successful.
+        Consensus::transfer(
+            ctx,
+            to,
+            &amount,
+            MessageEventHookInvocation::new(
+                CONSENSUS_TRANSFER_HANDLER.to_string(),
+                types::ConsensusTransferContext {
+                    address: to,
+                    amount: amount.clone(),
+                },
+            ),
+        )?;
+
+        Ok(())
+    }
+}
+
+impl<Accounts: modules::accounts::API, Consensus: modules::consensus::API>
+    Module<Accounts, Consensus>
+{
+    /// Deposit in the runtime.
+    fn tx_deposit(ctx: &mut TxContext<'_, '_>, body: types::Deposit) -> Result<(), Error> {
+        let signer = &ctx
+            .tx_auth_info()
+            .expect("should be called with a transaction ctx")
+            .signer_info[0];
+        let address = Address::from_pk(&signer.public_key);
+
+        Self::deposit(ctx, address, body.amount)
+    }
+
+    /// Withdraw from the runtime.
+    fn tx_withdraw(ctx: &mut TxContext<'_, '_>, body: types::Withdraw) -> Result<(), Error> {
+        // Signer.
+        let signer = &ctx
+            .tx_auth_info()
+            .expect("should be called with a transaction context")
+            .signer_info[0];
+        let address = Address::from_pk(&signer.public_key);
+
+        Self::withdraw(ctx, address, body.amount)
+    }
+
+    fn query_balance(
+        ctx: &mut DispatchContext<'_>,
+        args: types::BalanceQuery,
+    ) -> Result<types::AccountBalance, Error> {
+        let denomination = Consensus::consensus_denomination(ctx)?;
+        let balances = Accounts::get_balances(ctx.runtime_state(), args.addr)
+            .map_err(|_| Error::InvalidArgument)?;
+        let balance = balances
+            .balances
+            .get(&denomination)
+            .ok_or(Error::InvalidArgument)?;
+        Ok(types::AccountBalance {
+            balance: balance.clone(),
```

Comment:
Why is clone needed here?

Reply:
Cause the `balance` returned from balances map is a reference. Hm i guess i could replace the `get` with `remove` since we don't actually mutate the underlying storage here.